### PR TITLE
add mapping server handler

### PR DESCRIPTION
**Pull request by djasmine**
add mapping files server listing function
add mapping files server query
handle mapping files upload in front-end and server

### CODE REVIEW

**Comment by yubowenok** [comment]
I have to run now. Will check tmrw morning.

File: src/css/dialog.css [hunk 1]
```diff
@@ -0,0 +1,41 @@
+.modal-body td {
```
Comment (yubowenok) [inline src/css/dialog.css]: This file shouldn't be committed. You need to double-check your git usage to avoid similar things happening again.

**Comment by yubowenok** [comment]
Could you please add wiki for the query?

File: server/mapping.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * Lists all the mapping files.
+ * @param {string} mappingPath Path to the mapping file folder.
+ * @return {{
+ *   fileList: ?Array<string>
+ * }}
+ */
+mapping.query.list = function(mappingPath) {
+  return {
```
Comment (yubowenok) [inline server/mapping.js]: Why not returning the fileList object itself?

File: server/mapping.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+};
+
+/**
+ * Gets mapping rules.
+ * @param {string} filePath Path to the mapping file.
+ * @return {{mappingName: !Object<string>}}
+ * @private
+ */
+mapping.getMapping_ = function(filePath) {
+  var mappingName = {};
```
Comment (yubowenok) [inline server/mapping.js]: What the meaning of "mappingName"? I think "mapping" or "mappingRules" would be better. Besides you can return the object ifself.

File: src/dialog.js [hunk 4]
```diff
@@ -286,6 +286,13 @@ genotet.dialog.upload_ = function() {
       var btnUpload = modal.find('#btn-upload').prop('disabled', true);
       var btnFile = modal.find('#btn-file');
       var fileDisplay = modal.find('#file-display');
+
+      selectpicker.change(function() {
+        var isMapping = selectpicker.val() == 'mapping';
+        modal.find('#data-name').prop('disabled', isMapping);
```
Comment (yubowenok) [inline src/dialog.js]: You can hide the inputs instead of disable them.

**Comment by yubowenok** [comment]
Who is assigned to complete the UI that selects the mapping file? @djasmine @Shauro

**Comment by djasmine** [comment]
I think @Shauro will do the UI for selecting the mapping file. We will check it out.

**Comment by yubowenok** [comment]
Have you updated the wiki for mapping queries? @djasmine

File: server/server.js [hunk 5]
```diff
@@ -147,6 +156,9 @@ app.post('/genotet/upload', upload.single('file'), function(req, res) {
     case 'bed':
       prefix = bedPath;
       break;
+    case 'mapping':
```
Comment (yubowenok) [inline server/server.js:159]: Please add in readme.md to instruct user to add mappingPath in config.

File: src/dialog.js [hunk 6]
```diff
@@ -291,6 +291,19 @@ genotet.dialog.upload_ = function() {
       var btnUpload = modal.find('#btn-upload').prop('disabled', true);
       var btnFile = modal.find('#btn-file');
       var fileDisplay = modal.find('#file-display');
+
+      selectpicker.change(function() {
+        var isMapping = selectpicker.val() == 'mapping';
+        if (isMapping) {
+          console.log(modal.find('#data-name'));
+          modal.find('#data-name-row').prop('style', 'display:none');
```
Comment (yubowenok) [inline src/dialog.js]:
``` js
modal.find('#data-name-row').css('display', 'none');
```

Same for the below.

File: templates/upload.html [hunk 7]
```diff
@@ -16,18 +16,19 @@ <h4 class="modal-title">Upload Data</h4>
           <option value="binding">Binding Track</option>
           <option value="bed">Bed Track</option>
           <option value="expression">Expression Matrix</option>
+          <option value="mapping">Mapping</option>
         </select>
       </td>
     </tr>
-    <tr>
+    <tr id="data-name-row">
```
Comment (yubowenok) [inline templates/upload.html]: This is a bit too many ids. You can do 

``` js
modal.find('#data-name').closest('tr')
```

to retrive the tr container.

**Comment by djasmine** [comment]
Added wiki and readme.md. @yubowenok

**Comment by yubowenok** [comment]
LGTM

**Comment by djasmine** [comment]
There is still a front-end issue.
qq20160201-0 2x
The label at left side would contract when hiding some rows. How should I do with that. @yubowenok

**Comment by yubowenok** [comment]
That is because the width of the first table column is detemined by the longest string in that colume. Previously "description" is the longest. I think you can try to set the first column to not wrap on whitespaces. Try css "white-space: nowrap".

**Comment by djasmine** [comment]
Fixed the front-end issue. @yubowenok

**Comment by yubowenok** [comment]
already LGTMed. You may merge.

File: templates/upload.html [hunk 8]
```diff
@@ -9,13 +9,14 @@ <h4 class="modal-title">Upload Data</h4>
       <col width="100">
     </colgroup>
     <tr>
-      <td>Data Type</td>
+      <td style="white-space:nowrap">Data Type</td>
```
Comment (yubowenok) [inline templates/upload.html:12]: Please try to set this in css later. Inline css is to be avoided unless for initial property such as display none.

**Pull request merged**